File: architecturecomponents/src/main/java/com/fesskiev/architecturecomponents/domain/source/remote/RemoteDataSource.java
package com.fesskiev.architecturecomponents.domain.source.remote;


import android.content.Context;

import com.fesskiev.architecturecomponents.R;
import com.fesskiev.architecturecomponents.domain.entity.City;
import com.fesskiev.architecturecomponents.domain.entity.WeatherDesc;
import com.fesskiev.architecturecomponents.domain.entity.WeatherResponse;
import com.fesskiev.architecturecomponents.domain.source.DataSource;
import com.fesskiev.architecturecomponents.domain.source.remote.retrofit.APIService;

import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import io.reactivex.Observable;
import io.reactivex.Single;
import retrofit2.Retrofit;

public class RemoteDataSource implements DataSource {

    private APIService service;
    private Context context;

    public RemoteDataSource(Context context, Retrofit retrofit) {
        this.context = context;
        service = retrofit.create(APIService.class);
    }

    @Override
    public Single<WeatherResponse> getWeatherByCity(String city) {
        return service.getWeatherByCity(weatherQueryParams(city));
    }

    @Override
    public Observable<WeatherResponse> getWeatherByLocation(double latitude, double longitude) {
        return service.getWeatherByLocation(weatherLocationQueryParams(latitude, longitude));
    }

    private Map<String, String> weatherLocationQueryParams(double latitude, double longitude) {
        Map<String, String> data = new LinkedHashMap<>();
        data.put("lat", String.valueOf(latitude));
        data.put("lon", String.valueOf(longitude));
        data.put("mode", "json");
        data.put("units", "metric");
        data.put("cnt", "1");
        data.put("apikey", context.getString(R.string.open_weather_key));
        return data;
    }

    private Map<String, String> weatherQueryParams(String city) {
        Map<String, String> data = new LinkedHashMap<>();
        data.put("q", city);
        data.put("mode", "json");
        data.put("units", "metric");
        data.put("cnt", "7");
        data.put("apikey", context.getString(R.string.open_weather_key));
        return data;
    }

    @Override
    public void saveWeatherResponse(WeatherResponse weatherResponse) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Single<List<City>> getCities() {
        throw new UnsupportedOperationException();
    }

    @Override
    public Single<List<WeatherDesc>> getWeather(String cityId) {
        throw new UnsupportedOperationException();
    }

    @Override
    public Single<Object> deleteCity(City city) {
        throw new UnsupportedOperationException();
    }
}

File: kotlinsamples/src/main/java/com/fesskiev/kotlinsamples/domain/source/remote/RemoteDataSource.kt
package com.fesskiev.kotlinsamples.domain.source.remote

import android.content.Context
import com.fesskiev.kotlinsamples.R
import com.fesskiev.kotlinsamples.domain.entity.TopTracks
import com.fesskiev.kotlinsamples.domain.source.DataSource
import io.reactivex.Single
import retrofit2.Retrofit
import java.util.LinkedHashMap


class RemoteDataSource(private val context: Context, retrofit: Retrofit) : DataSource {

    private val service: ApiService = retrofit.create(ApiService::class.java)

    override fun getTopTracks(): Single<TopTracks> {
        return service.getTopTracks(topTracksParams())
    }

    private fun topTracksParams(): Map<String, String> {
        val data = LinkedHashMap<String, String>()
        data["method"] = "chart.gettoptracks"
        data["api_key"] = context.getString(R.string.last_fm_api_key)
        data["format"] = "json"
        return data
    }
}